Feat:  Implement the function in the market_utils library #3

Tbelleng opened the pull request:
<!--- Please provide a general summary of your changes in the title above -->

# Pull Request type

<!-- Please try to limit your pull request to one type; submit multiple pull requests if needed. -->

Please add the labels corresponding to the type of changes your PR introduces:

- Bugfix
- Feature

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Currently, functions are working but since last git pull i wasnt able to trigger tests
Resolves: #NA

## What is the new behavior?

<!-- Please describe the behavior or changes that are being added by this PR. -->

- Implement from getMaxPositionImpactFactor function to getExpectedMinTokenBalance in the market_utils library.
- Had to change some external files because functions prototypes or return values was not working 
- Add functions to convert int and uint.
- Some function are still missing from other files (the last function of market_utils #435 use some Keys functions that exist in solidity files but not in satoru files (i commented the line with what it should looks like)
- Commented "test_deposit_utils::given_unsufficient_fee_token_amount_for_deposit_then_fails" because one function essential function (get_market) was missing in Satoru. (Approved by @zarboq)

## Does this introduce a breaking change?

<!-- Yes or No -->
No
<!-- If this does introduce a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

<!-- Any other information that is important to this PR, such as screenshots of how the component looks before and after the change. -->

0xTrinityy commented:
- Tests passed
- Commented "test_deposit_utils::given_unsufficient_fee_token_amount_for_deposit_then_fails" because one function essential function (get_market) was missing in Satoru. (Approved by @zarboq)

delaaxe commented on src/market/market_utils.cairo:
```diff
@@ -57,6 +66,16 @@ struct GetNextFundingAmountPerSizeResult {
 }
 
+#[derive(Drop, starknet::Store, Serde)]
+struct GetExpectedMinTokenBalanceCache {
```
"Cache" structs are Solidity artifacts that shouldn't be needed in Cairo

delaaxe commented on src/market/market_utils.cairo:
```diff
@@ -1241,3 +1257,966 @@ fn get_adjusted_position_impact_factor(
+fn get_expected_min_token_balance(
+    data_store: IDataStoreDispatcher, market: Market, token: ContractAddress
+) -> u128 {
+    let mut cache = GetExpectedMinTokenBalanceCache {
```
You can just declare `let` variables as they appear instead of using this

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1660,33 +1700,47 @@ fn get_enabled_market(data_store: IDataStoreDispatcher, market_address: Contract
-    pnl_factor_type_for_shorts: felt252,
-) { //TODO
-}
+// fn validate_max_pnl(
```
remove comment

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -2274,105 +2191,962 @@ fn apply_delta_to_funding_fee_amount_per_size(
+    );
+
+    let is_exceeded: bool = pnl_to_pool_factor > 0
+        && pnl_to_pool_factor > from_u128_to_i128(max_pnl_factor);
```
```suggestion
        && cal::to_unsigned(pnl_to_pool_factor) > max_pnl_factor;
```

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -2274,105 +2191,962 @@ fn apply_delta_to_funding_fee_amount_per_size(
+    event_emitter.emit_ui_fee_factor_updated(account, ui_fee_factor);
+}
+
+fn validate_market_token_balance_add(data_store: IDataStoreDispatcher, markets: Array<Market>) {
```
```suggestion
fn validate_market_token_balance_array(data_store: IDataStoreDispatcher, markets: Array<Market>) {
```

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -2274,105 +2191,962 @@ fn apply_delta_to_funding_fee_amount_per_size(
+        if i == length {
+            break;
+        }
+        let index: Market = *markets.at(i);
+        validate_market_token_balance_check(data_store, index);
```
```suggestion
        validate_market_token_balance_check(data_store, *markets.at(i));
```

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -2274,105 +2191,962 @@ fn apply_delta_to_funding_fee_amount_per_size(
+    validate_market_token_balance_util(data_store, market, market.short_token);
+}
+
+fn validate_market_token_balance_util(
```
```suggestion
fn validate_market_token_balance_with_token(
```

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -2274,105 +2191,962 @@ fn apply_delta_to_funding_fee_amount_per_size(
+    };
+}
+
+fn validate_market_token_balance(data_store: IDataStoreDispatcher, market_add: ContractAddress) {
```
```suggestion
fn validate_market_address_token_balance(data_store: IDataStoreDispatcher, market_add: ContractAddress) {
```

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -2274,105 +2191,962 @@ fn apply_delta_to_funding_fee_amount_per_size(
+        market.market_token == 0.try_into().unwrap() || token == 0.try_into().unwrap(),
+        MarketError::EMPTY_ADDRESS_IN_MARKET_TOKEN_BALANCE_VALIDATION
+    );
+    let balancee: u256 = IERC20Dispatcher { contract_address: token }
```
```suggestion
    let balance: u256 = IERC20Dispatcher { contract_address: token }
```

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1573,6 +1591,32 @@ fn get_pnl_to_pool_factor(
     return get_pnl_to_pool_factor_from_prices(data_store, @market, @prices, is_long, maximize);
 }
 
+fn from_i128_to_u128(value: i128) -> u128 {
+    0
+}
```
remove, use calc::to_unsigned() instead

Tbelleng commented on src/market/market_utils.cairo:
```diff
@@ -2274,105 +2191,962 @@ fn apply_delta_to_funding_fee_amount_per_size(
+        if i == length {
+            break;
+        }
+        let market_adress_prev = swap_path.get(i);
+        let market_adress: ContractAddress = *market_adress_prev.unwrap().unbox();
```
This doesnt work apparently, getting error "Desnap operator can only be applied on snapshots" and "Ambiguous method call. More than one applicable trait function with a suitable self type was found"

zarboq replied:
Ok keep it like this then !

Tbelleng commented on src/market/market_utils.cairo:
```diff
@@ -2274,105 +2191,962 @@ fn apply_delta_to_funding_fee_amount_per_size(
+    let max_swap_path_length: u128 = data_store.get_u128(keys::max_swap_path_length());
+    let token_swap_path_length: u32 = token_swap_path.len();
+    assert(
+        token_swap_path_length <= from_u128_to_u32(max_swap_path_length),
```
No implementation for into given this context.

zarboq replied:
```suggestion
        token_swap_path_length.into() <= max_swap_path_length,
```

Tbelleng replied:
My bad, thanks Zarboq !

Tbelleng commented on src/market/market_utils.cairo:
```diff
@@ -2274,105 +2191,962 @@ fn apply_delta_to_funding_fee_amount_per_size(
+        if i == token_swap_path_length {
+            break;
+        }
+        let market_prev = token_swap_path.get(i);
+        let market: ContractAddress = *market_prev.unwrap().unbox();
```
got exactly the same error than above when merging line

zarboq replied:
keep it like it is then

Tbelleng commented on src/market/market_utils.cairo:
```diff
@@ -2274,105 +2191,962 @@ fn apply_delta_to_funding_fee_amount_per_size(
+    );
+    let claimable_collateral_amount: u128 = data_store
+        .get_u128(keys::claimable_collateral_amount_key(market.market_token, token));
+    let claimable_fee_amount: u128 = data_store
+        .get_u128(
+            keys::claimable_fee_amount()
+        ); // line must be :claimable_fee_amount_key(market.market_token, token));
+    let claimable_ui_fee_amount: u128 = data_store
+        .get_u128(
+            keys::claimable_ui_fee_amount()
+        ); //line must be : claimable_ui_fee_amount_key(market.market_token, token));
```
Because it didn't exist at the time I did it, but it has been fixed now

Tbelleng commented:
I did the error function for MAX_SWAP_PATH_LENGTH_EXCEEDED, but for an unknown reason it failed the test "tests::exchange::test_withdrawal_handler::given_normal_conditions_when_create_withdrawal_then_works".

its return the following failed test "[FAIL] tests::exchange::test_withdrawal_handler::given_normal_conditions_when_create_withdrawal_then_works

Failure data:
    original value: [2948898997531897529390896139788207291365863622618688546988264006116708], converted to a string: [max_swap_path_length_exceeded]
    original value: [0], converted to a string: []
    original value: [0], converted to a string: []"

Any hint about it ?

zarboq commented:
It's bc you inverted the inequality, it should be `if token_swap_path_length.into() > max_swap_path_length` @Tbelleng

Tbelleng commented:
My bad ! thanks @zarboq , it should works now ⚡️

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1591,19 +1591,7 @@ fn get_pnl_to_pool_factor(
-}
-
-fn from_u256_to_u128(value: u256) -> u128 {
+fn get_u256_low(value: u256) -> u128 {
```
No need for a function you can juste do `value.low` where it's needed

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -2965,7 +2939,9 @@ fn validate_max_pnl(
     );
 
-    assert(!is_pnl_factor_exceeded_for_longs, MarketError::PNL_EXCEEDED_FOR_LONGS);
+    if (!is_pnl_factor_exceeded_for_longs) {
```
```suggestion
    if (is_pnl_factor_exceeded_for_longs) {
```

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -2974,7 +2950,9 @@ fn validate_max_pnl(
     );
 
-    assert(!is_pnl_factor_exceeded_for_shorts, MarketError::PNL_EXCEEDED_FOR_SHORTS);
+    if (!is_pnl_factor_exceeded_for_shorts) {
```
```suggestion
    if (is_pnl_factor_exceeded_for_shorts) {
```

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -3041,50 +3019,53 @@ fn set_ui_fee_factor(
     let max_ui_fee_factor: u128 = data_store.get_u128(keys::max_ui_fee_factor());
 
-    assert(ui_fee_factor <= max_ui_fee_factor, MarketError::UI_FEE_FACTOR_EXCEEDED);
+    if (ui_fee_factor <= max_ui_fee_factor) {
```
`if (ui_fee_factor > max_ui_fee_factor) {`

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -3099,20 +3080,20 @@ fn validate_market_token_balance_util(
-        balance >= collateral_amount,
-        MarketError::INVALID_MARKET_TOKEN_BALANCE_FOR_COLLATERAL_AMOUNT
-    );
+    if (balance >= collateral_amount) {
```
```suggestion
    if (balance < collateral_amount) {
```

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -3099,20 +3080,20 @@ fn validate_market_token_balance_util(
-        balance >= claimable_funding_fee_amount,
-        MarketError::INVALID_MARKET_TOKEN_BALANCE_FOR_CLAIMABLE_FUNDING
-    );
+    if (balance >= claimable_funding_fee_amount) {
```
```suggestion
    if (balance < claimable_funding_fee_amount) {
```

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1535,17 +1546,24 @@ fn validate_open_interest(data_store: IDataStoreDispatcher, market: @Market, is_
     }
 }
 
+/// Validata the swap market.
+/// # Arguments
+/// * `data_store` - The data store to use.
+/// * `market` - The market to validate the open interest for.
+// fn validate_swap_market(data_store: @IDataStoreDispatcher, market: @Market) { // TODO
+// }
+
```
remove comment

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1535,17 +1546,24 @@ fn validate_open_interest(data_store: IDataStoreDispatcher, market: @Market, is_
-    // TODO
-    0
-}
+// fn get_min_pnl_factor_after_adl(
+//     data_store: IDataStoreDispatcher, market: ContractAddress, is_long: bool
+// ) -> u128 {
+//     // TODO
+//     0
+// }
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1573,6 +1591,20 @@ fn get_pnl_to_pool_factor(
+    0
+}
+
+// Get the ratio of pnl to pool value.
+// # Arguments
+// * `data_store` - The data_store dispatcher.
+// * `market` Rhe market.
+// * `prices` the prices of the market tokens.
+// * `is_long` whether to get the value for the long or short side.
+// * `maximize` whether to maximize the factor.
+// # Returns
+// (pnl of positions) / (long or short pool value)
+// TODO same function names getPnlToPoolFactor
 /// Get the ratio of PNL (Profit and Loss) to pool value.
 ///
 /// # Arguments
```
Same comment twice

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1607,16 +1639,16 @@ fn get_pnl_to_pool_factor_from_prices(
-    // TODO
-    (true, 0, 0)
-}
+// fn is_pnl_factor_exceeded(
+//     data_store: IDataStoreDispatcher,
+//     oracle: IOracleDispatcher,
+//     market_address: ContractAddress,
+//     is_long: bool,
+//     pnl_factor_type: felt252
+// ) -> (bool, u128, u128) {
+//     // TODO
+//     (true, 0, 0)
+// }
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1635,21 +1667,17 @@ fn is_pnl_factor_exceeded_direct(
-        max_ui_fee_factor
-    }
-}
+// fn get_ui_fee_factor(data_store: IDataStoreDispatcher, account: ContractAddress) -> u128 {
+//     let max_ui_fee_factor = data_store.get_u128(keys::max_ui_fee_factor());
+//     let ui_fee_factor = data_store.get_u128(keys::ui_fee_factor_key(account));
+//     if ui_fee_factor < max_ui_fee_factor {
+//         ui_fee_factor
+//     } else {
+//         max_ui_fee_factor
+//     }
+// }
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1635,21 +1667,17 @@ fn is_pnl_factor_exceeded_direct(
-/// * `dataStore` - DataStore
-/// * `marketAddress` - The address of the market.
-fn get_enabled_market(data_store: IDataStoreDispatcher, market_address: ContractAddress) -> Market {
+fn get_market(data_store: IDataStoreDispatcher, market_address: ContractAddress) -> Market {
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1660,33 +1688,32 @@ fn get_enabled_market(data_store: IDataStoreDispatcher, market_address: Contract
 }
 
 
+/// Gets the enabled market. This function will revert if the market does not exist or is not enabled.
+/// # Arguments
+/// * `dataStore` - DataStore
+/// * `marketAddress` - The address of the market.
+// fn get_enabled_market(data_store: IDataStoreDispatcher, market_address: ContractAddress) -> Market {
+//     //TODO
+//     Market {
+//         market_token: Zeroable::zero(),
+//         index_token: Zeroable::zero(),
+//         long_token: Zeroable::zero(),
+//         short_token: Zeroable::zero(),
+//     }
+// }
+
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1660,33 +1688,32 @@ fn get_enabled_market(data_store: IDataStoreDispatcher, market_address: Contract
-    pnl_factor_type_for_shorts: felt252,
-) { //TODO
-}
+// fn get_cumulative_borrowing_factor(
+//     data_store: @IDataStoreDispatcher, market: ContractAddress, is_long: bool
+// ) -> u128 {
+//     (*data_store).get_u128(keys::cumulative_borrowing_factor_key(market, is_long))
+// }
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1697,8 +1724,8 @@ fn validate_market_token_balance_with_address(
 
-fn validate_market_token_balance(data_store: IDataStoreDispatcher, market: Market) { //TODO
-}
+// fn validate_market_token_balance(data_store: IDataStoreDispatcher, market: Market) { //TODO
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1707,35 +1734,34 @@ fn validate_markets_token_balance(data_store: IDataStoreDispatcher, market: Span
 /// * `data_store`: dispatcher for the data store
 /// * `market`: the market to check
-fn validate_position_market(data_store: IDataStoreDispatcher, market: Market) {} // TODO
+// fn validate_position_market(data_store: IDataStoreDispatcher, market: Market) {} // TODO
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1707,35 +1734,34 @@ fn validate_markets_token_balance(data_store: IDataStoreDispatcher, market: Span
-) -> Array<Market> { //TODO
-    Default::default()
-}
+// fn get_swap_path_markets(
+//     data_store: IDataStoreDispatcher, swap_path: Span32<ContractAddress>
+// ) -> Array<Market> { //TODO
+//     Default::default()
+// }
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1707,35 +1734,34 @@ fn validate_markets_token_balance(data_store: IDataStoreDispatcher, market: Span
-    };
-}
-
+// fn (data_store: IDataStoreDispatcher, market: Market) {
+//     assert(!market.market_token.is_zero(), MarketError::EMPTY_MARKET);
+//     let is_market_disabled = data_store.get_bool(keys::is_market_disabled_key(market.market_token));
+
+//     match is_market_disabled {
+//         Option::Some(result) => {
+//             assert(!result, MarketError::DISABLED_MARKET);
+//         },
+//         Option::None => {
+//             panic_with_felt252(MarketError::DISABLED_MARKET);
+//         }
+//     };
+// }
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1750,48 +1776,27 @@ fn validate_enabled_market_address(
-fn is_market_collateral_token(market: Market, token: ContractAddress) -> bool {
-    token == market.long_token || token == market.short_token
-}
+// fn is_market_collateral_token(market: Market, token: ContractAddress) -> bool {
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1750,48 +1776,27 @@ fn validate_enabled_market_address(
-    0
-}
-
+// fn validate_market_collateral_token(market: Market, token: ContractAddress) {
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1750,48 +1776,27 @@ fn validate_enabled_market_address(
 /// * `open_interest_delta` - The change in open interest.
 /// * `is_long` - Whether it is for the long or short side
-fn get_min_collateral_factor_for_open_interest(
+fn _for_open_interest(
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1834,16 +1839,16 @@ fn update_cumulative_borrowing_factor(
-    next_position_borrowing_factor: u128
-) { // TODO
-}
+// fn update_total_borrowing(
+//     data_store: IDataStoreDispatcher,
+//     market: ContractAddress,
+//     is_long: bool,
+//     prev_position_size_in_usd: u128,
+//     prev_position_borrowing_factor: u128,
+//     next_position_size_in_usd: u128,
+//     next_position_borrowing_factor: u128
+// ) { // TODO
+// }
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1852,11 +1857,11 @@ fn update_total_borrowing(
-) -> u128 { // TODO
-    0
-}
+// fn market_token_amount_to_usd(
+//     market_token_amount: u128, pool_value: u128, supply: u128
+// ) -> u128 { // TODO
+//     0
+// }
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -2234,145 +2136,993 @@ fn get_virtual_inventory_for_swaps(
+// the total amount of position collateral for a market
+fn get_collateral_sum(
+    data_store: IDataStoreDispatcher,
+    market: Market,
```
should be ContractAddress

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -2234,145 +2136,993 @@ fn get_virtual_inventory_for_swaps(
+
+    // use 1 for the getCollateralSum divisor since getCollateralSum does not sum over both the
+    // longToken and shortToken
+    let mut collateral_amount: u128 = get_collateral_sum(data_store, market, token, true, 1);
+    collateral_amount += get_collateral_sum(data_store, market, token, false, 1);
```
You should fix the function param for market to take an address and use `market.market_token`

delaaxe commented on src/market/market_utils.cairo:
```diff
@@ -2234,145 +1935,995 @@ fn get_virtual_inventory_for_swaps(
+        market.market_token.is_non_zero() && token.is_non_zero(),
+        MarketError::EMPTY_ADDRESS_IN_MARKET_TOKEN_BALANCE_VALIDATION
+    );
+    let balance: u128 = IERC20Dispatcher { contract_address: token }
```
better `try_into` and fail if value doesn't fit

delaaxe commented on src/event/event_emitter.cairo:
```diff
@@ -559,11 +559,15 @@ trait IEventEmitter<TContractState> {
+    fn emit_ui_fee_factor_updated_check(
+        ref self: TContractState, account: ContractAddress, ui_fee_factor: u128
+    );
+
```
why is this duplicated function needed?

delaaxe commented on src/market/market_utils.cairo:
```diff
@@ -2,13 +2,18 @@
+use result::ResultTrait;
+
+use debug::PrintTrait;
+use zeroable::Zeroable;
```
last 3 imports unnecessary

delaaxe commented on src/market/market_utils.cairo:
```diff
@@ -17,20 +22,23 @@ use satoru::market::{
+use satoru::utils::precision::{apply_exponent_factor, float_to_wei, mul_div};
+use satoru::data::keys::{skip_borrowing_fee_for_smaller_side, max_swap_path_length};
+// use satoru::utils::arrays::get_u128;
+//use satoru::utils::{calc, precision};
```
commented code

delaaxe commented on src/market/market_utils.cairo:
```diff
@@ -1318,14 +1323,20 @@ fn get_swap_impact_amount_with_cap(
-fn get_open_interest(
+fn get_open_interest(data_store: IDataStoreDispatcher, market: Market, is_long: bool,) -> u128 {
+    0
+}
```
shouldn't this be implemented?

zarboq replied:
I think it's `get_open_interest_for_market_is_long` but it's this empty implementation that gets called so we should remove it

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1444,10 +1455,10 @@ fn get_pool_divisor(long_token: ContractAddress, short_token: ContractAddress) -
-    data_store: IDataStoreDispatcher, token_swap_path: Span32<ContractAddress>
-) { //TODO
-}
+// fn validate_swap_path(
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1697,46 +1623,9 @@ fn validate_market_token_balance_with_address(
 ) { //TODO
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1697,46 +1623,9 @@ fn validate_market_token_balance_with_address(
-fn validate_market_token_balance(data_store: IDataStoreDispatcher, market: Market) { //TODO
-}
-
 fn validate_markets_token_balance(data_store: IDataStoreDispatcher, market: Span<Market>) { //TODO
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1746,59 +1635,6 @@ fn validate_enabled_market_address(
 ) { // TODO
```
remove

zarboq commented on src/market/market_utils.cairo:
```diff
@@ -1635,58 +1613,6 @@ fn is_pnl_factor_exceeded_direct(
     (true, 0, 0)
```
remove function

Tbelleng commented:
solved issues but after removing "get_open_interest" as requested, i got this error when compiling "error: WithdrawalHandler: failed to compile Sierra contract to CASM

Caused by:
    #35998: One of the arguments does not satisfy the requirements of the libfunc." did i do something wrong ?

zarboq commented:
Do this in `get_borrowing_factor_per_second`

```
let market_snap = @market;
if (skip_borrowing_fee_for_smaller_side) {
    let long_open_interest: u128 = get_open_interest_for_market_is_long(data_store, market_snap, true);
    let short_open_interest: u128 = get_open_interest_for_market_is_long(data_store, market_snap, false);

    // if getting the borrowing factor for longs and if the longOpenInterest
    // is smaller than the shortOpenInterest, then return zero
    if (is_long && long_open_interest < short_open_interest) {
        return 0;
    }
    // if getting the borrowing factor for shorts and if the shortOpenInterest
    // is smaller than the longOpenInterest, then return zero
    if (!is_long && short_open_interest < long_open_interest) {
        return 0;
    }
}
```
